Importing Libraries

In [147]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

sns.set_style('darkgrid')
matplotlib.rcParams['font.size'] = 14
matplotlib.rcParams['figure.figsize'] = (9, 5)
matplotlib.rcParams['figure.facecolor'] = '#00000000'

In [148]:
import warnings
warnings.filterwarnings("ignore")

Accessing Data Files

In [149]:
IPL_matches_df = pd.read_csv(R"C:\Users\heman\OneDrive\Documents\IPL Data Analysis\IPL_Matches_2008_2022.csv")

Accessing the first 5 Data Values

In [150]:
IPL_matches_df.head()

ID       City        Date Season  MatchNumber  \
0  1312200  Ahmedabad  2022-05-29   2022        Final   
1  1312199  Ahmedabad  2022-05-27   2022  Qualifier 2   
2  1312198    Kolkata  2022-05-25   2022   Eliminator   
3  1312197    Kolkata  2022-05-24   2022  Qualifier 1   
4  1304116     Mumbai  2022-05-22   2022           70   

                         Team1                 Team2  \
0             Rajasthan Royals        Gujarat Titans   
1  Royal Challengers Bangalore      Rajasthan Royals   
2  Royal Challengers Bangalore  Lucknow Super Giants   
3             Rajasthan Royals        Gujarat Titans   
4          Sunrisers Hyderabad          Punjab Kings   

                              Venue            TossWinner TossDecision  \
0  Narendra Modi Stadium, Ahmedabad      Rajasthan Royals          bat   
1  Narendra Modi Stadium, Ahmedabad      Rajasthan Royals        field   
2             Eden Gardens, Kolkata  Lucknow Super Giants        field   
3             Eden Gardens, Kolkata        Gujarat Titans        field   
4          Wankhede Stadium, Mumbai   Sunrisers Hyderabad          bat   

  SuperOver                  WinningTeam    WonBy  Margin method  \
0         N               Gujarat Titans  Wickets     7.0    NaN   
1         N             Rajasthan Royals  Wickets     7.0    NaN   
2         N  Royal Challengers Bangalore     Runs    14.0    NaN   
3         N               Gujarat Titans  Wickets     7.0    NaN   
4         N                 Punjab Kings  Wickets     5.0    NaN   

  Player_of_Match                                       Team1Players  \
0       HH Pandya  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
1      JC Buttler  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
2      RM Patidar  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
3       DA Miller  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
4   Harpreet Brar  ['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...   

                                        Team2Players        Umpire1  \
0  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...    CB Gaffaney   
1  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...    CB Gaffaney   
2  ['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...  J Madanagopal   
3  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...   BNJ Oxenford   
4  ['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...   AK Chaudhary   

         Umpire2  
0    Nitin Menon  
1    Nitin Menon  
2       MA Gough  
3      VK Sharma  
4  NA Patwardhan

Accessing the bottom 5 Data values

In [151]:
IPL_matches_df.tail()

ID        City        Date   Season MatchNumber  \
945  335986     Kolkata  2008-04-20  2007/08           4   
946  335985      Mumbai  2008-04-20  2007/08           5   
947  335984       Delhi  2008-04-19  2007/08           3   
948  335983  Chandigarh  2008-04-19  2007/08           2   
949  335982   Bangalore  2008-04-18  2007/08           1   

                           Team1                        Team2  \
945        Kolkata Knight Riders              Deccan Chargers   
946               Mumbai Indians  Royal Challengers Bangalore   
947             Delhi Daredevils             Rajasthan Royals   
948              Kings XI Punjab          Chennai Super Kings   
949  Royal Challengers Bangalore        Kolkata Knight Riders   

                                          Venue                   TossWinner  \
945                                Eden Gardens              Deccan Chargers   
946                            Wankhede Stadium               Mumbai Indians   
947                            Feroz Shah Kotla             Rajasthan Royals   
948  Punjab Cricket Association Stadium, Mohali          Chennai Super Kings   
949                       M Chinnaswamy Stadium  Royal Challengers Bangalore   

    TossDecision SuperOver                  WinningTeam    WonBy  Margin  \
945          bat         N        Kolkata Knight Riders  Wickets     5.0   
946          bat         N  Royal Challengers Bangalore  Wickets     5.0   
947          bat         N             Delhi Daredevils  Wickets     9.0   
948          bat         N          Chennai Super Kings     Runs    33.0   
949        field         N        Kolkata Knight Riders     Runs   140.0   

    method Player_of_Match                                       Team1Players  \
945    NaN       DJ Hussey  ['WP Saha', 'BB McCullum', 'RT Ponting', 'SC G...   
946    NaN      MV Boucher  ['L Ronchi', 'ST Jayasuriya', 'DJ Thornely', '...   
947    NaN     MF Maharoof  ['G Gambhir', 'V Sehwag', 'S Dhawan', 'MK Tiwa...   
948    NaN      MEK Hussey  ['K Goel', 'JR Hopes', 'KC Sangakkara', 'Yuvra...   
949    NaN     BB McCullum  ['R Dravid', 'W Jaffer', 'V Kohli', 'JH Kallis...   

                                          Team2Players    Umpire1  \
945  ['AC Gilchrist', 'Y Venugopal Rao', 'VVS Laxma...  BF Bowden   
946  ['S Chanderpaul', 'R Dravid', 'LRPL Taylor', '...   SJ Davis   
947  ['T Kohli', 'YK Pathan', 'SR Watson', 'M Kaif'...  Aleem Dar   
948  ['PA Patel', 'ML Hayden', 'MEK Hussey', 'MS Dh...  MR Benson   
949  ['SC Ganguly', 'BB McCullum', 'RT Ponting', 'D...  Asad Rauf   

            Umpire2  
945     K Hariharan  
946       DJ Harper  
947  GA Pratapkumar  
948      SL Shastri  
949     RE Koertzen

To see whole DataFrame we have to set max row greater than our DataFrame rows.

In [152]:
pd.options.display.max_rows = 999

Finding out the no.of ROWS and COLUMNS

In [153]:
IPL_matches_df.shape

(950, 20)

There is 950 rows and 20 columns in our Dataset

Accessing the Data COLUMNS

In [154]:
IPL_matches_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 950 entries, 0 to 949
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               950 non-null    int64  
 1   City             899 non-null    object 
 2   Date             950 non-null    object 
 3   Season           950 non-null    object 
 4   MatchNumber      950 non-null    object 
 5   Team1            950 non-null    object 
 6   Team2            950 non-null    object 
 7   Venue            950 non-null    object 
 8   TossWinner       950 non-null    object 
 9   TossDecision     950 non-null    object 
 10  SuperOver        946 non-null    object 
 11  WinningTeam      946 non-null    object 
 12  WonBy            950 non-null    object 
 13  Margin           932 non-null    float64
 14  method           19 non-null     object 
 15  Player_of_Match  946 non-null    object 
 16  Team1Players     950 non-null    object 
 17  Team2Players    

Accessing the Data sets

In [155]:
IPL_matches_df.nunique()

ID                 950
City                33
Date               706
Season              15
MatchNumber         80
Team1               18
Team2               18
Venue               49
TossWinner          18
TossDecision         2
SuperOver            2
WinningTeam         18
WonBy                4
Margin              94
method               1
Player_of_Match    262
Team1Players       932
Team2Players       937
Umpire1             58
Umpire2             60
dtype: int64

Finding for empty cells

In [156]:
IPL_matches_df.isnull().values.any()

True

In [157]:
IPL_matches_df.isnull().sum()

ID                   0
City                51
Date                 0
Season               0
MatchNumber          0
Team1                0
Team2                0
Venue                0
TossWinner           0
TossDecision         0
SuperOver            4
WinningTeam          4
WonBy                0
Margin              18
method             931
Player_of_Match      4
Team1Players         0
Team2Players         0
Umpire1              0
Umpire2              0
dtype: int64

In [158]:
IPL_matches_df.isnull().sum().sum()

1012

There are a total of 1012 Empty cells in our data set

Data Cleaning

In [159]:
IPL_matches_df.loc[(IPL_matches_df['Venue']== 'Dubai International Cricket Stadium'),['City']] = 'Dubai'
IPL_matches_df.loc[(IPL_matches_df['Venue'] == 'Sharjah Cricket Stadium'),['City']] = 'Sharjah'
IPL_matches_df['SuperOver'].fillna('NoResult',inplace=True)
IPL_matches_df['WinningTeam'].fillna('NoResult',inplace=True)
IPL_matches_df['Player_of_Match'].fillna('NoResult',inplace= True)
IPL_matches_df['Margin'].fillna('NoMargin',inplace = True)

In [160]:
IPL_matches_df=IPL_matches_df.drop(['method'],axis =1 )

In [161]:
IPL_matches_df.isnull().any()

ID                 False
City               False
Date               False
Season             False
MatchNumber        False
Team1              False
Team2              False
Venue              False
TossWinner         False
TossDecision       False
SuperOver          False
WinningTeam        False
WonBy              False
Margin             False
Player_of_Match    False
Team1Players       False
Team2Players       False
Umpire1            False
Umpire2            False
dtype: bool

Removing Duplicates

In [162]:
IPL_matches_df.duplicated().any()

False

Renaming the teams who changed their names after some seasons

In [163]:
IPL_matches_df['Team1']=IPL_matches_df['Team1'].str.replace('Kings XI Punjab','Punjab Kings')
IPL_matches_df['Team2']=IPL_matches_df['Team2'].str.replace('Kings XI Punjab','Punjab Kings')
IPL_matches_df['WinningTeam']=IPL_matches_df['WinningTeam'].str.replace('Kings XI Punjab','Punjab Kings')

IPL_matches_df['Team1']=IPL_matches_df['Team1'].str.replace('Delhi Daredevils','Delhi Capitals')
IPL_matches_df['Team2']=IPL_matches_df['Team2'].str.replace('Delhi Daredevils','Delhi Capitals')
IPL_matches_df['WinningTeam']=IPL_matches_df['WinningTeam'].str.replace('Delhi Daredevils','Delhi Capitals')

IPL_matches_df['Team1']=IPL_matches_df['Team1'].str.replace('Rising Pune Supergiants','Rising Pune Supergiant')
IPL_matches_df['Team2']=IPL_matches_df['Team2'].str.replace('Rising Pune Supergiants','Rising Pune Supergiant')
IPL_matches_df['WinningTeam'] = IPL_matches_df['WinningTeam'].str.replace('Rising Pune Supergiants', 'Rising Pune Supergiant')

In [164]:
IPL_matches_df.head()

ID       City        Date Season  MatchNumber  \
0  1312200  Ahmedabad  2022-05-29   2022        Final   
1  1312199  Ahmedabad  2022-05-27   2022  Qualifier 2   
2  1312198    Kolkata  2022-05-25   2022   Eliminator   
3  1312197    Kolkata  2022-05-24   2022  Qualifier 1   
4  1304116     Mumbai  2022-05-22   2022           70   

                         Team1                 Team2  \
0             Rajasthan Royals        Gujarat Titans   
1  Royal Challengers Bangalore      Rajasthan Royals   
2  Royal Challengers Bangalore  Lucknow Super Giants   
3             Rajasthan Royals        Gujarat Titans   
4          Sunrisers Hyderabad          Punjab Kings   

                              Venue            TossWinner TossDecision  \
0  Narendra Modi Stadium, Ahmedabad      Rajasthan Royals          bat   
1  Narendra Modi Stadium, Ahmedabad      Rajasthan Royals        field   
2             Eden Gardens, Kolkata  Lucknow Super Giants        field   
3             Eden Gardens, Kolkata        Gujarat Titans        field   
4          Wankhede Stadium, Mumbai   Sunrisers Hyderabad          bat   

  SuperOver                  WinningTeam    WonBy Margin Player_of_Match  \
0         N               Gujarat Titans  Wickets    7.0       HH Pandya   
1         N             Rajasthan Royals  Wickets    7.0      JC Buttler   
2         N  Royal Challengers Bangalore     Runs   14.0      RM Patidar   
3         N               Gujarat Titans  Wickets    7.0       DA Miller   
4         N                 Punjab Kings  Wickets    5.0   Harpreet Brar   

                                        Team1Players  \
0  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
1  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
2  ['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...   
3  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...   
4  ['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...   

                                        Team2Players        Umpire1  \
0  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...    CB Gaffaney   
1  ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...    CB Gaffaney   
2  ['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...  J Madanagopal   
3  ['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...   BNJ Oxenford   
4  ['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...   AK Chaudhary   

         Umpire2  
0    Nitin Menon  
1    Nitin Menon  
2       MA Gough  
3      VK Sharma  
4  NA Patwardhan

Accessing a new Data set

In [165]:
IPL_ball_by_ball_df = pd.read_csv(R"C:\Users\heman\OneDrive\Documents\IPL Data Analysis\IPL_Ball_by_Ball_2008_2022.csv")
IPL_ball_by_ball_df.head()

ID  innings  overs  ballnumber       batter          bowler  \
0  1312200        1      0           1  YBK Jaiswal  Mohammed Shami   
1  1312200        1      0           2  YBK Jaiswal  Mohammed Shami   
2  1312200        1      0           3   JC Buttler  Mohammed Shami   
3  1312200        1      0           4  YBK Jaiswal  Mohammed Shami   
4  1312200        1      0           5  YBK Jaiswal  Mohammed Shami   

   non-striker extra_type  batsman_run  extras_run  total_run  non_boundary  \
0   JC Buttler        NaN            0           0          0             0   
1   JC Buttler    legbyes            0           1          1             0   
2  YBK Jaiswal        NaN            1           0          1             0   
3   JC Buttler        NaN            0           0          0             0   
4   JC Buttler        NaN            0           0          0             0   

   isWicketDelivery player_out kind fielders_involved       BattingTeam  
0                 0        NaN  NaN               NaN  Rajasthan Royals  
1                 0        NaN  NaN               NaN  Rajasthan Royals  
2                 0        NaN  NaN               NaN  Rajasthan Royals  
3                 0        NaN  NaN               NaN  Rajasthan Royals  
4                 0        NaN  NaN               NaN  Rajasthan Royals

In [166]:
IPL_ball_by_ball_df.shape

(225954, 17)

Data Cleaning

In [167]:
IPL_ball_by_ball_df['extra_type'].fillna('N',inplace = True)
IPL_ball_by_ball_df['player_out'].fillna('N',inplace = True)
IPL_ball_by_ball_df['kind'].fillna('N',inplace = True)
IPL_ball_by_ball_df['fielders_involved'].fillna('N',inplace = True)

In [168]:
IPL_ball_by_ball_df.isnull().values.any()

False

Data Analysis

Teams that have played in the last 15 seasons

In [169]:
IPL_matches_df.Team1.unique().reshape(15,1)

array([['Rajasthan Royals'],
       ['Royal Challengers Bangalore'],
       ['Sunrisers Hyderabad'],
       ['Delhi Capitals'],
       ['Chennai Super Kings'],
       ['Gujarat Titans'],
       ['Lucknow Super Giants'],
       ['Kolkata Knight Riders'],
       ['Punjab Kings'],
       ['Mumbai Indians'],
       ['Rising Pune Supergiant'],
       ['Gujarat Lions'],
       ['Pune Warriors'],
       ['Deccan Chargers'],
       ['Kochi Tuskers Kerala']], dtype=object)

Total matches played by each team

In [174]:
match_df = pd.concat([IPL_matches_df['Team1'], IPL_matches_df['Team2']]).value_counts().reset_index()
match_df.set_axis(['Teams','Total_Matches'],axis = 'columns')
match_df

index  count
0                Mumbai Indians    231
1   Royal Challengers Bangalore    226
2                Delhi Capitals    224
3         Kolkata Knight Riders    223
4                  Punjab Kings    218
5           Chennai Super Kings    208
6              Rajasthan Royals    192
7           Sunrisers Hyderabad    152
8               Deccan Chargers     75
9                 Pune Warriors     46
10       Rising Pune Supergiant     30
11                Gujarat Lions     30
12               Gujarat Titans     16
13         Lucknow Super Giants     15
14         Kochi Tuskers Kerala     14

In [175]:
def get_color(name, number):
    pal = list(sns.color_palette(palette=name, n_colors=number).as_hex())
    return pal

In [176]:
pal_vi = get_color('viridis_r', len(match_df))

In [179]:
%pip install plotly --upgrade --quiet
import plotly.express as px
fig = px.bar(match_df, x='Teams', y='Total_Matches', text='Total_Matches',
             color ='Teams', color_discrete_sequence=pal_vi)

fig.update_traces(texttemplate='%{text:.3s}', textposition='outside')
fig.update_layout({'plot_bgcolor': 'white',
                   'paper_bgcolor': 'white'})

fig.update_layout(width=1100, height=500,
                  margin = dict(t=15, l=15, r=15, b=15))
fig.show()

^C
Note: you may need to restart the kernel to use updated packages.


ValueError: Value of 'x' is not the name of a column in 'data_frame'. Expected one of ['index', 'count'] but received: Teams